source/common/formatter/substitution_formatter.cc

```diff
@@ -464,7 +466,8 @@ SubstitutionFormatParser::parse(const std::string& format,
                                 const std::vector<CommandParserPtr>& commands) {
   std::string current_token;
   std::vector<FormatterProviderPtr> formatters;
-  const std::regex command_w_args_regex(R"EOF(^%([A-Z]|[0-9]|_)+(\([^\)]*\))?(:[0-9]+)?(%))EOF");
+  const std::regex command_w_args_regex(
+      R"EOF(^%([A-Z]|[0-9]|_)+(\([^\)]*\))?(:[0-9]+)?(%)|^%%)EOF");
```

Comment:
I don't think this regular expression should match `%%` because it forces every instance of %% to add another formatter object. Instead, I think below, between the check for `!= '%'` and the check for a non-empty `current_token`, you could just see if the next character is also `%` (making sure to double-check bounds). If you find a double-%, append a single `'%'` to `current_token`, increment pos, and continue the loop.

This makes sure that the `%%` sequence is included in any preceding or subsequent PlainStringFormatter, and we can keep that regex focused on actual commands, don't need to add an EscapeFormatter type, and so on.